High level lambdify documentation

asmeurer opened the pull request:
**This is ready to review.**

I have completely rewritten the `lambdify` docstring. Of note is a new section describing how lambdify works. I believe if someone understands the basic idea of how it works as a namespace translation, then they will be less likely to fall into some of the common gotchas with the function, such as mixing SymPy and NumPy types. I have also removed some parts that were not helpful. 

This also contains some minor fixes to the tests module:

- The `doctest_depends_on` decorator now supports `python_version`, which can be used to skip doctests in Python 2. 
- `doctest -v` now shows the test name for `doctest_depends_on` tests. 

Some of the examples purport to show the type of the output, e.g., as a NumPy array, but it isn't actually clear. This will be better once https://github.com/sympy/sympy/issues/13483 is fixed, because then NumPy arrays with print with `array(...)` and mpmath types with print with `mpf` in the doctests.

Fixes #13482.
Fixes #13483 

<!-- BEGIN RELEASE NOTES -->
* utilities
  * added high-level documentation to the docstring of `lambdify` to help clarify what it does and how it works.

asmeurer committed:
Fix doctest_depends_on for lambidfy()

There are still some errors in the doctests, which I do not know how to fix.
There appears to be some bug in the tensorflow code. Also, for some reasons,
NumPy arrays are not printing with the "array(...)" part.

Related to #13483 (but doesn't completely fix it).

asmeurer committed:
Add some better high level documentation to lambdify

This still needs some cleaning up. I'm unsure exactly how the different
sections should be ordered. Also, I'm unsure if this makes the docstring too
long, and if some parts should be moved to Sphinx. I also didn't yet test the
Sphinx rendering :)

Fixes #13482.

moorepants commented:
NIce addition. Thanks.

It is a bit long for a docstring. Maybe a module level docstring or on a page like this:

http://docs.sympy.org/latest/modules/codegen.html

In the long run I think we need a clear division between the tutorial, prose documentation (like this PR), and docstrings. The scikit-learn docs are really nicely organized: http://scikit-learn.org/stable/

asmeurer commented:
I'm also unsure about the whole "accidentally working" gotcha section. It almost seems to me that we should just fix lambdify to be better in cases like that. It does come up a lot, though, so if it isn't changed I feel it should be documented somehow.

asmeurer commented:
I can't figure out what the Sphinx error is referring to.

asmeurer committed:
Small formatting improvements to the lambdify docstring

ylemkimon commented on sympy/utilities/lambdify.py:
```diff
@@ -176,12 +231,202 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
+    NumPy), and only pass it input types that are compatible with that module
+    (say, NumPy arrays).** Remember that by default, if the ``module``
+    argument is not provided, ``lambdify`` creates functions using the NumPy
+    namespace.
```
I think it should mention it uses "math", "mpmath", "sympy" if NumPy is not installed, here too.

ylemkimon commented:
Also there is #13326.

bjodah reviewed (approved):
Sorry for the late reply. Forgot about this and stumbled upon it now. I think it looks good, I agree with @moorepants that most of the text perhaps belong in the docs outside lambdify's docstring (what about at the module level of lambdify.py ?)

asmeurer committed:
Merge branch 'master' into lambdify-doc

asmeurer committed:
Update the lambdify docstring with the latest changes to lambdify

lambdify now creates real functions, not lambda functions.

I have also added a bit showing how to inspect the source code and namespace
of a lambdified function.

asmeurer committed:
Fix the doctests of lambdify

Note that currently the doctester doesn't properly use repr() for numpy
arrays, so the output is not ideal. This is a separate issue that needs to be
fixed. See #13483.

asmeurer committed:
Some more cleanups of the lambdify docstring

Move things around a bit to make the docstring more coherent, and remove some
bits that were not very helpful, especially parts that suggested mixing of
modules.

asmeurer committed:
lambda is a keyword, not a builtin

asmeurer committed:
Fix Sphinx build issues in the lambdify docstring

asmeurer committed:
Fix Sphinx rendering of some examples in the lambdify docstring

asmeurer committed:
Remove the automatic reordering of the "notes" and "examples" sections in Sphinx

There is no special formatting applied to these sections (we do not use the
plot feature of numpydoc), so there is no reason to treat them as special
headers. It's annoying to have these sections in a specific place in the
docstring and for Sphinx to move them around. For example, if you put "notes"
at the end, without this patch, it would move it to the top.

It still does reorder other sections, such as see also and references, but
these have special formatting applied to them, so it's less of an issue.
Potentially moving to napoleon would fix this (I haven't 100% confirmed). See
issue #15375.

asmeurer committed:
Remove nested formatting

RST doesn't support nested formatting.

asmeurer committed:
Merge branch 'sphinx-reordering' into HEAD

asmeurer committed:
Add python_version support to doctest_depends_on

asmeurer committed:
Print the name of doctest_depends_on skipped tests in doctest -v

asmeurer committed:
Skip the lambdify doctests in Python 2

They use exec, which has different syntax in Python 2. Trying to make it
compatible with both would unnecessarily complicate the example. Since we
won't be supporting Python 2 for very long anyway, it is simplest to just skip
the doctests in Python 2 altogether.

asmeurer committed:
Fix the lambdify doctests

oscarbenjamin commented:
Looks like the test failure is because the printer identifies the Tensor class from tensorflow with the Tensor class from Sympy: it only checks the class name before assuming that a `_print_Tensor` method exists.

asmeurer committed:
Fix the lambdify tensorflow doctest

The doctest fails because it tries to use the SymPy printers to print the
tensorflow object, which fails because it uses the printer for the SymPy
Tensor object.

As a workaround, just print it directly in the doctest.

asmeurer commented:
I guess this is also related to the `repr` issue noted at https://github.com/sympy/sympy/issues/13483.

oscarbenjamin commented:
I don't see how to reproduce it outside the doctests:
```julia
In [1]:     >>> import tensorflow as tf 
   ...:     >>> from sympy import Max 
   ...:     >>> f = Max(x, sin(x)) 
   ...:     >>> func = lambdify(x, f, 'tensorflow') 
   ...:     >>> result = func(tf.constant(1.0)) 
   ...:     >>> result # a tf.Tensor representing the result of the calculation                                                                
/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: compiletime version 3.6 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.7
  return f(*args, **kwds)
Out[1]: <tf.Tensor 'Maximum:0' shape=() dtype=float32>

In [2]: result                                                                                                                                 
Out[2]: <tf.Tensor 'Maximum:0' shape=() dtype=float32>

In [3]: init_printing(pretty=False)                                                                                                            

In [4]: result                                                                                                                                 
Out[4]: <tf.Tensor 'Maximum:0' shape=() dtype=float32>

In [5]: from sympy.utilities.runtests import setup_pprint                                                                                      

In [6]: setup_pprint()                                                                                                                         
Out[6]: True

In [7]: result                                                                                                                                 
Out[7]: <tf.Tensor 'Maximum:0' shape=() dtype=float32>
```

The same error (`AttributeError: _print_Tensor`) shows up in the doctests under both pytest and bin/test.

asmeurer committed:
Add a note that MPMATH_TRANSLATIONS is used in Function._eval_evalf

See https://github.com/sympy/sympy/issues/15795#issuecomment-455384608.

asmeurer committed:
More cleanups to the lambdify docstring

asmeurer commented:
OK, I have cleaned this up and it is now ready for reviewing. 

I have also merged in https://github.com/sympy/sympy/pull/15828. If that is an issue, I can remove it. It was too annoying having the notes section, which should go at the end, be put at the beginning. 

Note that there are several docstring examples here that demonstrate the type of something such as a numpy array or mpmath float, which do not actually do that, because the doctester does not use `repr` on non-SymPy types (see https://github.com/sympy/sympy/issues/13483). I have left the doctests as they are, in the hopes that that issue will be solved and the doctests will make sense again.  (To fix it, one needs to take https://github.com/asmeurer/sympy/tree/repr-doctests and fix all the doctests. This is an easy fix for anyone interested).

asmeurer commented:
By the way, a helpful tip for anyone working on the Sphinx docs. The following temporary change will make the docs build a lot faster

```diff
diff --git a/doc/src/conf.py b/doc/src/conf.py
index d972ff7cbb..854e9e5d4e 100644
--- a/doc/src/conf.py
+++ b/doc/src/conf.py
@@ -22,7 +22,7 @@

 # Add any Sphinx extension module names here, as strings. They can be extensions
 # coming with Sphinx (named 'sphinx.addons.*') or your custom ones.
-extensions = ['sphinx.ext.autodoc', 'sphinx.ext.viewcode',
+extensions = ['sphinx.ext.autodoc', #'sphinx.ext.viewcode',
               'sphinx.ext.mathjax', 'numpydoc', 'sympylive',
               'sphinx.ext.graphviz', 'matplotlib.sphinxext.plot_directive']
```

I wish the viewcode extension would use the same caching as the rest of Sphinx.

oscarbenjamin commented:
@asmeurer  I don't think that your branch at [https://github.com/asmeurer/sympy/tree/repr-doctests](https://github.com/asmeurer/sympy/tree/repr-doctests) does the right thing. I just tested with:
```console
$ pytest --doctest-plus --doctest-rst -k 'not test_' -n4
```

Here are some example failures:
```
110     >>> x = 'abc'
111     >>> expr = x + 'def'
112     >>> expr
Expected:
    'abcdef'
Got:
    abcdef

268     >>> classify_pde(eq)
Expected:
    ('1st_linear_constant_coeff_homogeneous',)
Got:
    (1st_linear_constant_coeff_homogeneous,)

115     >>> from sympy import *
116     >>> x, y, z = symbols('x y z')
117     >>> str(Integral(sqrt(1/x), x))
Expected:
    'Integral(sqrt(1/x), x)'
Got:
    Integral(sqrt(1/x), x)
```

It probably needs to copy the logic from [https://github.com/asmeurer/sympy/blob/8cbe07f8d44fd828c967927a8e97df689b18081f/sympy/printing/str.py#L38](https://github.com/asmeurer/sympy/blob/8cbe07f8d44fd828c967927a8e97df689b18081f/sympy/printing/str.py#L38)

asmeurer commented:
I think the printer my custom printer subclasses from also defines printers for str and tuple. So those need to be overridden as well.

asmeurer commented:
I pushed a change to `repr-doctests` that should do that. A better fix might be to make doprint not even dispatch to non-SymPy types generically.

oscarbenjamin commented:
The new patch in `repr-doctests` has increased the number of fails from 191 to 309. I haven't worked through them all but a quick glance shows that they are mostly down to ordering:
```
Expected:
    {x, y}
Got:
    {y, x}
```
This is precisely one of the issues I fixed in getting the doctests to work with pytest: we need the StrPrinter for this.

asmeurer commented:
Oh I forgot about the sorting. So it's not as straightforward as my patch. But still should be fixable.

asmeurer commented:
Actually can we just remove the `_print_list`, `_print_dict`, `_print_tuple`, and `_print_set` redefinitions?

oscarbenjamin commented:
```diff
diff --git a/sympy/printing/str.py b/sympy/printing/str.py
index 406efbb62..ba9448da9 100644
--- a/sympy/printing/str.py
+++ b/sympy/printing/str.py
@@ -858,8 +858,7 @@ class ReprPrinter(StrPrinter):
     def emptyPrinter(self, expr):
         return repr(expr)
 
-    _print_str = _print_list = _print_tuple = _print_dict = _print_set = \
-    _print_int = _print_float = _print_mpq = emptyPrinter
+    _print_str = _print_int = _print_float = _print_mpq = emptyPrinter
 
 def sstrrepr(expr, **settings):
     """return expr in mixed str/repr form
```

Now there are 112 fails. They now seem to be examples where repr differs from str as expected e.g.:
```
Expected:
    [[1]
     [2]]
Got:
    matrix([[1],
            [2]])

Expected:
    Old feature has been deprecated since SymPy 1.1. Use new feature
    instead. See https://github.com/sympy/sympy/issues/5241 for more info.
Got:
    SymPyDeprecationWarning(None, 'Old feature', None, 'new feature', 5241, '1.1')
```

There are also a lot of these in polys:
```
Expected:
    [x**3 - 2*x**2 + x - 3, 3*x**2 - 4*x + 1, 2/9*x + 25/9, -2079/4]
Got:
    [x**3 - mpq(2,1)*x**2 + x - mpq(-3,1), mpq(3,1)*x**2 - mpq(4,1)*x + mpq(1,1), mpq(2,9)*x + mpq(25,9), -mpq(-2079,4)]

Expected:
    (1/2*x + 1, 5)
Got:
    (mpq(1,2)*x + mpq(1,1), mpq(5,1))
```
Do those want to be fixed? Maybe the printer could also handle mpq etc.

asmeurer commented:
We should try to match the regular Python shell as much as possible. For me, it doesn't show mpq in the regular Python shell when I run that example. 

Actually for mpq we have to handle it anyway because gmpy won't be used when it isn't installed, meaning the doctests would fail.

asmeurer commented:
By the way, we may eventually have to consider the fact that dicts are automatically sorted in Python 3.6+. That likely won't matter until down the road when we drop 3.5 support ([in late 2020](https://github.com/sympy/sympy/wiki/Python-version-support-policy)).

oscarbenjamin commented:
Having a reliable ordering for doctests etc. could make some things a bit easier. As long as SymPy needs to have a custom printer for doctests then ordering sets/dicts isn't a big deal though.

oscarbenjamin commented:
In fact using this
```diff
$ git diff
diff --git a/sympy/printing/str.py b/sympy/printing/str.py
index 406efbb62..b1305dde2 100644
--- a/sympy/printing/str.py
+++ b/sympy/printing/str.py
@@ -858,8 +858,7 @@ class ReprPrinter(StrPrinter):
     def emptyPrinter(self, expr):
         return repr(expr)
 
-    _print_str = _print_list = _print_tuple = _print_dict = _print_set = \
-    _print_int = _print_float = _print_mpq = emptyPrinter
+    _print_str = _print_int = _print_float = emptyPrinter
 
 def sstrrepr(expr, **settings):
     """return expr in mixed str/repr form
```

I only get 31 fails. Examples:
```
132     >>> from sympy.polys.rings import sring
133     >>> R, expr = sring(1/a**3 + a**QQ(3, 7)); R
Expected:
    Polynomial ring in 1/a, a**(1/7) over ZZ with lex order
Got:
    Polynomial ring in 1/a, a**(1/7) over ZZ with LexOrder() order

218     >>> from sympy import lambdify, Matrix
219     >>> from sympy.abc import x, y
220     >>> import numpy
221     >>> array2mat = [{'ImmutableDenseMatrix': numpy.matrix}, 'numpy']
222     >>> f = lambdify((x, y), Matrix([x, y]), modules=array2mat)
223     >>> f(1, 2)
Expected:
    [[1]
     [2]]
Got:
    matrix([[1],
            [2]])

044     >>> SymPyDeprecationWarning(
Expected:
    Old feature has been deprecated since SymPy 1.1. Use new feature
    instead. See https://github.com/sympy/sympy/issues/5241 for more info.
Got:
    SymPyDeprecationWarning(None, 'Old feature', None, 'new feature', 5241, '1.1')

090     >>> from sympy.unify.core import unify, Compound, Variable
091     >>> expr    = Compound("Add", ("x", "y"))
092     >>> pattern = Compound("Add", ("x", Variable("a")))
093     >>> next(unify(expr, pattern, {}))
Expected:
    {Variable(a): 'y'}
Got:
    {<sympy.unify.core.Variable object at 0x1243d93c8>: 'y'

```

asmeurer commented:
The first one of those (`R, expr = sring(1/a**3 + a**QQ(3, 7)); R`) does not match what you get in a regular Python shell. The others do.

asmeurer commented:
Thanks @smichr!
